File: js/web/lib/wasm/jsep/webgpu/ops/batch-norm.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import {TensorView} from '../../tensor-view';
import {ShapeUtil} from '../../util';
import {AttributeWithCacheKey, createAttributeWithCacheKey} from '../attribute-with-cache-key';
import {ComputeContext, ProgramInfo} from '../types';

import {inputVariable, outputVariable, ShaderHelper} from './common';

export interface BatchNormAttributes extends AttributeWithCacheKey {
  readonly epsilon: number;
  readonly momentum: number;
  readonly spatial: boolean;
  readonly trainingMode: boolean;
  readonly format: 'nhwc'|'nchw';
  readonly outputCount: number;
}

const validateInputs = (inputs: readonly TensorView[], attributes: BatchNormAttributes): void => {
  if (!inputs || inputs.length !== 5) {
    throw new Error('BatchNormalization requires 5 inputs');
  }

  const checkShapeEqual = (actual: readonly number[], expected: readonly number[], message: string) => {
    const r = expected.length;
    if (r !== actual.length) {
      throw new Error(`${message}: num dimensions != ${r}`);
    }
    expected.forEach((v, i) => {
      if (v !== actual[i]) {
        switch (i % 100) {
          case 11:
          case 12:
          case 13:
            throw new Error(`${message}: ${i}th dimension != ${v}`);
          default:
        }
        switch (i % 10) {
          case 1:
            throw new Error(`${message}: ${i}st dimension != ${v}`);
          case 2:
            throw new Error(`${message}: ${i}nd dimension != ${v}`);
          case 3:
            throw new Error(`${message}: ${i}rd dimension != ${v}`);
          default:
            throw new Error(`${message}: ${i}th dimension != ${v}`);
        }
      }
    });
  };

  if (inputs[0].dims.length > 1) {
    const shape = Object.seal<Record<typeof attributes.format, number[]>>({
      nhwc: inputs[0].dims.slice(-1),
      nchw: inputs[0].dims.slice(1, attributes.spatial ? 2 : undefined),
    })[attributes.format];
    checkShapeEqual(inputs[1].dims, shape, 'Invalid input scale');
    checkShapeEqual(inputs[2].dims, shape, 'Invalid input B');
    checkShapeEqual(inputs[3].dims, shape, 'Invalid input mean');
    checkShapeEqual(inputs[4].dims, shape, 'Invalid input var');
  } else {
    checkShapeEqual(inputs[1].dims, [1], 'Invalid input scale');
    checkShapeEqual(inputs[2].dims, [1], 'Invalid input B');
    checkShapeEqual(inputs[3].dims, [1], 'Invalid input mean');
    checkShapeEqual(inputs[4].dims, [1], 'Invalid input var');
  }
};

const createBatchNormInferenceProgramInfo = (inputs: readonly TensorView[], attributes: BatchNormAttributes):
    ProgramInfo => {
      const {epsilon, spatial, format} = attributes;
      const yShape = inputs[0].dims;
      const outputSize = ShapeUtil.size(yShape);
      const x = inputVariable('x', inputs[0].dataType, inputs[0].dims);
      const scale = inputVariable('scale', inputs[1].dataType, [ShapeUtil.size(inputs[1].dims)]);
      const bias = inputVariable('bias', inputs[2].dataType, [ShapeUtil.size(inputs[2].dims)]);
      const inputMean = inputVariable('inputMean', inputs[3].dataType, [ShapeUtil.size(inputs[3].dims)])
      const inputVar = inputVariable('inputVar', inputs[4].dataType, [ShapeUtil.size(inputs[4].dims)]);
      const y = outputVariable('y', inputs[0].dataType, yShape);

      const calcCOffset = (): string => {
        let cOffset = '';
        if (spatial) {
          cOffset = `let cOffset = ${
              yShape.length === 1   ? '0u' :
                  format === 'nhwc' ? `outputIndices[${yShape.length - 1}]` :
                                      `outputIndices[1]`};`;
        } else {
          if (format === 'nchw') {
            cOffset = `
            ${y.indicesSet('outputIndices', '0', '0')}
            let cOffset = ${y.indicesToOffset('outputIndices')};`;
          } else {
            // update C channel.
            cOffset = `var cIndices = ${scale.type.indices}('0');
                       cIndices[0] = outputIndices[${yShape.length - 1}];`;
            // update D1 x ... x Dn channels.
            for (let i = 1; i < scale.rank; i++) {
              cOffset += `cIndices[${i}] = outputIndices[${i + 1}];`;
            }
            cOffset += `let cOffset = ${scale.indicesToOffset('cIndices')};`;
          }
        }
        return cOffset;
      };
      const getInferenceModeShaderSource = (helper: ShaderHelper) => `
  const epsilon = ${epsilon};
  ${helper.declareVariables(x, scale, bias, inputMean, inputVar, y)}
  ${helper.mainStart()}
  ${helper.guardAgainstOutOfBoundsWorkgroupSizes(outputSize)}
    var outputIndices = ${y.offsetToIndices('global_idx')};
    ${calcCOffset()}
    let scale = ${scale.getByOffset('cOffset')};
    let bias = ${bias.getByOffset('cOffset')};
    let inputMean = ${inputMean.getByOffset('cOffset')};
    let inputVar = ${inputVar.getByOffset('cOffset')};
    let x = ${x.getByOffset('global_idx')};
    let value = (x - inputMean) / sqrt(inputVar + epsilon) * scale + bias;
    ${y.setByOffset('global_idx', 'value')}
  }`;
      return {
        name: 'BatchNormalization',
        shaderCache: {hint: `${attributes.epsilon}_${attributes.format}_${spatial}`},
        getShaderSource: getInferenceModeShaderSource,
        getRunData: () => ({
          outputs: [{dims: inputs[0].dims, dataType: inputs[0].dataType}],
          dispatchGroup: {x: Math.ceil(outputSize / 64 /* workgroup size */)},
        }),
      };
    }

export const parseBatchNormAttributes = (attributes: Record<string, unknown>): BatchNormAttributes =>
    createAttributeWithCacheKey(attributes as Omit<BatchNormAttributes, keyof AttributeWithCacheKey>);

export const batchNorm = (context: ComputeContext, attributes: Record<string, unknown>): void => {
  const {inputs, outputCount} = context;
  const updatedAttributes = parseBatchNormAttributes({...attributes, outputCount});
  validateInputs(inputs, updatedAttributes);
  if (attributes.trainingMode) {
    throw new Error('BatchNormalization trainingMode is not supported yet.');
  } else {
    context.compute(createBatchNormInferenceProgramInfo(inputs, updatedAttributes));
  }
};

File: onnxruntime/core/providers/webgpu/nn/batch_norm.cc
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "core/common/inlined_containers.h"
#include "core/providers/webgpu/nn/batch_norm.h"
#include "core/providers/cpu/nn/batch_norm_helper.h"
#include "core/providers/cpu/tensor/utils.h"
#include "core/providers/webgpu/shader_helper.h"
#include "core/providers/webgpu/webgpu_supported_types.h"

namespace onnxruntime {
namespace webgpu {

#define WEBGPU_BATCH_NORM_VERSIONED_KERNEL(start, end, domain, is_nhwc) \
  ONNX_OPERATOR_VERSIONED_KERNEL_EX(                                    \
      BatchNormalization,                                               \
      domain,                                                           \
      start,                                                            \
      end,                                                              \
      kWebGpuExecutionProvider,                                         \
      (*KernelDefBuilder::Create())                                     \
          .TypeConstraint("T", WebGpuSupportedFloatTypes()),            \
      BatchNormalization<is_nhwc>);

#define WEBGPU_BATCH_NORM_KERNEL(version, domain, is_nhwc)   \
  ONNX_OPERATOR_KERNEL_EX(                                   \
      BatchNormalization,                                    \
      domain,                                                \
      version,                                               \
      kWebGpuExecutionProvider,                              \
      (*KernelDefBuilder::Create())                          \
          .TypeConstraint("T", WebGpuSupportedFloatTypes()), \
      BatchNormalization<is_nhwc>);

WEBGPU_BATCH_NORM_VERSIONED_KERNEL(7, 8, kOnnxDomain, false)
WEBGPU_BATCH_NORM_VERSIONED_KERNEL(9, 13, kOnnxDomain, false)
WEBGPU_BATCH_NORM_VERSIONED_KERNEL(14, 14, kOnnxDomain, false)
WEBGPU_BATCH_NORM_KERNEL(15, kOnnxDomain, false)

WEBGPU_BATCH_NORM_VERSIONED_KERNEL(7, 8, kMSInternalNHWCDomain, true)
WEBGPU_BATCH_NORM_VERSIONED_KERNEL(9, 13, kMSInternalNHWCDomain, true)
WEBGPU_BATCH_NORM_VERSIONED_KERNEL(14, 14, kMSInternalNHWCDomain, true)
WEBGPU_BATCH_NORM_KERNEL(15, kMSInternalNHWCDomain, true)

Status BatchNormalizationProgram::GenerateShaderCode(ShaderHelper& shader) const {
  const ShaderVariableHelper& input_tensor = shader.AddInput("input_tensor", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  const ShaderVariableHelper& scale = shader.AddInput("scale", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  const ShaderVariableHelper& B = shader.AddInput("B", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  const ShaderVariableHelper& input_mean = shader.AddInput("input_mean", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  const ShaderVariableHelper& input_var = shader.AddInput("input_var", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);
  const ShaderVariableHelper& output = shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias);

  shader.MainFunctionBody() << shader.GuardAgainstOutOfBoundsWorkgroupSizes("uniforms.output_size")
                            << "  let idx = global_idx * " << components_ << ";\n"
                            << "  var outputIndices = " << output.OffsetToIndices("idx") << ";\n";
  if (spatial_) {
    if (input_tensor.Rank() == 1) {
      shader.MainFunctionBody() << "  let cOffset = 0u;\n";
    } else {
      if (format_ == DataLayout::NHWC) {
        shader.MainFunctionBody() << "  let cOffset = outputIndices[" << input_tensor.Rank() - 1 << "] / " << components_ << ";\n";
      } else {
        shader.MainFunctionBody() << "  let cOffset = outputIndices[1];\n";
      }
    }
  } else {
    if (format_ == DataLayout::NCHW) {
      shader.MainFunctionBody() << "  " << output.IndicesSet("outputIndices", "0", "0") << "\n"
                                << "  let cOffset = " << output.IndicesToOffset("outputIndices") << ";\n";
    } else {
      // update C channel
      shader.MainFunctionBody() << "  var cIndices = scale_indices_t(0);\n"
                                << "  cIndices[0] = outputIndices[" << input_tensor.Rank() - 1 << "];\n";
      // update D1 x ... x Dn channels
      for (int i = 1; i < scale.Rank(); i++) {
        shader.MainFunctionBody() << "  cIndices[" << i << "] = outputIndices[" << i << "];\n";
      }
      shader.MainFunctionBody() << "  let cOffset = " << scale.IndicesToOffset("cIndices") << ";\n";
    }
  }

  shader.MainFunctionBody() << "  let scale = " << scale.GetByOffset("cOffset") << ";\n"
                            << "  let B = " << B.GetByOffset("cOffset") << ";\n"
                            << "  let input_mean = " << input_mean.GetByOffset("cOffset") << ";\n"
                            << "  let input_var = " << input_var.GetByOffset("cOffset") << ";\n"
                            << "  let x = " << input_tensor.GetByOffset("global_idx") << ";\n"
                            << "  let value = (x - input_mean) * inverseSqrt(input_var + " << epsilon_ << ") * scale + B;\n"
                            << "  " << output.SetByOffset("global_idx", "value") << "\n";

  return Status::OK();
}

template <bool is_nhwc>
Status BatchNormalization<is_nhwc>::ComputeInternal(ComputeContext& context) const {
  if (training_mode_) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "BatchNormalization trainingMode is not supported yet.");
  }

  if (context.InputCount() != 5) {
    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "BatchNormalization requires 5 inputs.");
  }

  const auto* input_tensor = context.Input(0);
  const TensorShape& input_shape = input_tensor->Shape();
  size_t input_rank = input_shape.NumDimensions();
  const int components = spatial_ ? ((input_shape[input_rank - 1] % 4 == 0) ? 4 : ((input_shape[input_rank - 1] % 2 == 0) ? 2 : 1)) : 1;

  auto output_dims = input_shape.AsShapeVector();
  TensorShape output_shape(output_dims);
  auto* output_tensor = context.Output(0, output_shape);
  int64_t output_size = output_tensor->Shape().Size() / static_cast<int64_t>(components);

  if (output_size == 0) {
    return Status::OK();
  }

  const auto* scale = context.Input<Tensor>(1);
  const auto* B = context.Input<Tensor>(2);
  const auto* input_mean = context.Input<Tensor>(3);
  const auto* input_var = context.Input<Tensor>(4);

  ORT_RETURN_IF_ERROR(BatchNormHelper::ValidateInputs(input_tensor, scale, B, input_mean, input_var, spatial_ == 1, format_ == DataLayout::NHWC));

  BatchNormalizationProgram program{epsilon_, spatial_, format_, static_cast<int64_t>(components)};
  program
      .AddInputs({{input_tensor, ProgramTensorMetadataDependency::TypeAndRank},
                  {scale, ProgramTensorMetadataDependency::TypeAndRank},
                  {B, ProgramTensorMetadataDependency::TypeAndRank},
                  {input_mean, ProgramTensorMetadataDependency::TypeAndRank},
                  {input_var, ProgramTensorMetadataDependency::TypeAndRank}})
      .AddOutputs({output_tensor})
      .SetDispatchGroupSize((output_size + WORKGROUP_SIZE - 1) / WORKGROUP_SIZE)
      .AddUniformVariables({{static_cast<uint32_t>(output_size)}});
  return context.RunProgram(program);
}

}  // namespace webgpu
}  // namespace onnxruntime